### PR TITLE
Remove thiserror dependency

### PR DESCRIPTION
**Pull request by GuillaumeGomez**

### CODE REVIEW

File: cairo/src/error.rs [hunk 1]
```diff
@@ -1,143 +1,156 @@
...
-    #[error("LastStatus")]
     #[doc(alias = "STATUS_LAST_STATUS")]
     LastStatus,
-    #[error("Unknown {0}")]
     #[doc(hidden)]
     __Unknown(i32),
 }
+
+impl std::error::Error for Error {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
```
Comment (sdroege) [inline cairo/src/error.rs:99]: This is the default implementation already

File: gio/src/input_stream.rs [hunk 2]
```diff
@@ -525,12 +526,27 @@ impl<T: IsA<InputStream>> InputStreamAsyncBufRead<T> {
...
+#[derive(Debug)]
 enum BufReadError {
-    #[error("Previous read operation failed")]
     Failed,
 }
 
+impl std::error::Error for BufReadError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        match self {
+            Self::Failed => None,
```
Comment (sdroege) [inline gio/src/input_stream.rs:537]: This makes no sense, that's the default implementation

**Review by sdroege — COMMENTED** [review]
So much boilerplate and much harder to review code. The original was obviously correct, yours is not.

**Comment by GuillaumeGomez** [comment]
> The original was obviously correct, yours is not.

Which part is invalid?

File: gio/src/input_stream.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply gio/src/input_stream.rs:537]: No, this is needed.

File: cairo/src/error.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply cairo/src/error.rs:99]: No, this is needed, or I misunderstood what you meant.

File: gio/src/input_stream.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply gio/src/input_stream.rs]: Ok nevermind, I didn't realize I removed the `derive(Error)`...

**Comment by GuillaumeGomez** [comment]
Apparently not, we have to implement `std::error::Error` by hand.

File: cairo/src/error.rs (hunk 1 above)
Comment (bilelmoussaoui) [reply cairo/src/error.rs:99]: You don't have to provide the source method if you're going to return None. That's the default implementation

Comment (sdroege) [reply cairo/src/error.rs:99]: Apparently implementing `Error` is complicated enough that you missed this once when implementing it and a second time after I pointed it out :shrug:

**Comment by sdroege** [comment]
> Which part is invalid?

I don't know, I didn't see any further errors that you introduced but the point is that it's absolutely obvious if there's anything wrong in the `thiserror` usage while you now added some 100 lines of code that has to be carefully reviewed.

**Comment by Hofer-Julian** [comment]
@GuillaumeGomez did you measure how much faster this crate compiles with your branch?

File: cairo/src/error.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply cairo/src/error.rs:99]: Sorry...

**Comment by GuillaumeGomez** [comment]
> @GuillaumeGomez did you measure how much faster this crate compiles with your branch?

A few seconds.

**Comment by sdroege** [comment]
> A few seconds.

You're not measuring correctly then. We still depend on `thiserror` indirectly, check your `Cargo.lock`.

39.91s for a clean build from master, 41.03s with your branch for a clean build here. In theory your branch should be minimally faster, so I guess they're both equally fast and any difference is in the margin of measurement errors.

**Comment by GuillaumeGomez** [comment]
It took 2 min for me, so 3 secs is in the margin.

**Comment by sdroege** [comment]
Yeah my point is that a) we still depend on `thiserror` anyway so you still compile it, b) if you try often enough both builds and average you'll see that they're both approximately the same. This has no effect other than adding more code.

**Comment by GuillaumeGomez** [comment]
Absolutely. However this is more readable and I know what's happening just by reading it unlike `thiserror`.

**Comment by sdroege** [comment]
It's absolutely not more readable. It's just a lot of repetitive code where some silly mistake can easily be hidden :)

**Comment by GuillaumeGomez** [comment]
Let's close it for now then...

**Pull request closed without merging**